src/interop.ts

```diff
@@ -12,25 +12,304 @@
+    constructor(provider: PulumiProvider) {
+        if (provider === PulumiProvider.AWS_NATIVE) {
+            this.pulumiMetadata = this.readMetadata();
+        } else {
+            throw new Error('AWS_NATIVE is the only supported pulumi provider');
+        }
+    }
+
+    /**
+     * Finds a specific resource in the metadata. Throws an UnknownCfnType error
+     * if one is not found (probably means it is a classic type).
+     *
+     * @param cfnType The CloudFormation type i.e. AWS::S3::Bucket
+     * @param metadata The PulumiMetadata to search
+     * @returns The PulumiResource
+     * @throws UnknownCfnType if the resource is not found
+     */
+    public findResource(cfnType: string): PulumiResource {
+        const pType = typeToken(cfnType);
+        if (pType in this.pulumiMetadata.resources) {
+            return this.pulumiMetadata.resources[pType];
+        }
+        throw new UnknownCfnType();
+    }
+
+    public types(): { [key: string]: PulumiType } {
+        return this.pulumiMetadata.types;
+    }
+
+    /**
+     * Read the aws-native metadata.json file
+     *
+     * @returns the PulumiMetadata
+     */
+    private readMetadata(): PulumiMetadata {
+        const contents = fs.readFileSync(path.join(__dirname, '../schemas/aws-native-metadata.json'), {
+            encoding: 'utf-8',
+        });
+        const metadata: PulumiMetadata = JSON.parse(contents);
+        return metadata;
+    }
+}
+
+/**
+ * If a property is a JSON type then we need provide
+ * the value as is, without further processing.
+ */
+enum NativeType {
+    /**
+     * The type is a json type and should be left as is
+     */
+    JSON = 'JSON',
+    /**
+     * The type is not a json type and should be processed
+     */
+    NON_JSON = 'NON_JSON',
+}
+
+/**
+ * Pulumi metadata.json schema. The real schema is a lot more detailed,
+ * but we only need pieces of the schema for our purposes
+ */
+export interface PulumiMetadata {
+    /**
+     * The Pulumi resource schemas
+     */
+    resources: { [key: string]: PulumiResource };
+
+    /**
+     * The Pulumi types
+     */
+    types: { [key: string]: PulumiType };
+}
+
+/**
+ * The Pulumi types. The type schema has a lot more attributes,
+ * but for our purposes we only need the properties
+ */
+interface PulumiType {
+    /**
+     * The properties of a type
+     */
+    properties: { [key: string]: PulumiProperty };
+}
+
+/**
+ * The schema for the array items property
+ */
+interface PulumiPropertyItems {
+    /**
+     * A reference to another type
+     */
+    $ref?: string;
+
+    /**
+     * The simple type (i.e. string, number, etc)
+     */
+    type?: string;
+}
+
+/**
+ * The schema for an individual property on a resource
+ */
+interface PulumiProperty extends PulumiPropertyItems {
+    /**
+     * If the property is an array then it will have an items property
+     */
+    items?: PulumiPropertyItems;
+}
+
+interface PulumiResource {
```

Comment:
Is this info coming from a Pulumi Package Schema, implicitly? E.g. that's the format with "$ref" references and such? 

I spent some time thinking about how this fits here, we're in a rather unfortunate situation that all nicer-to-use models for handling Pulumi Package Schema information are written in Go. This code looks like it's starting to build one in TS. Perhaps we just have to do this?

What could be an alternative? I guess we could do some more static analysis in Go at build-time and pre-compute some easy-to-use runtime metadata for TS. Does't necessarily sound very attractive to me, but listing as an option.

If we go with formalizing Pulumi Package Schema in TS I think we could lean harder in TS typesystem to make code airtight type-safe, but we can start easy and iterate on this.

Reply:
> Is this info coming from a Pulumi Package Schema, implicitly? E.g. that's the format with "$ref" references and such?

Yes.

> If we go with formalizing Pulumi Package Schema in TS I think we could lean harder in TS typesystem to make code airtight type-safe, but we can start easy and iterate on this.

I think it would be nice to have this in all languages that pulumi supports building component libraries in. For my purposes I only need some of the information from the schema so I think it should be out of scope to build the complete typescript schema.